devmapper: Disable mount option "discard" by default

rhvgoyal opened the pull request:
Fixes issue #12902 

Right now devicemapper mounts thin device using online discards by default
and passes mount option "discard". Generally people discourage usage of
online discards as they can be a drain on performance. Instead it is 
recommended to use fstrim once in a while to reclaim the space.

In case of  containers, we recommend to keep data volumes separate. So
there might not be lot of rm, unlink operations going on and there might
not be lot of space being freed by containers. So it might not matter
much if we don't reclaim that free space in pool.

User can still pass mount option explicitly using dm.mountopt=discard to
enable discards if they would like to.

So this is more like setting the containers by default for better performance
instead of better space efficiency in pool. And user can change the behavior
if they don't like default behavior.

Signed-off-by: Vivek Goyal vgoyal@redhat.com

rhvgoyal commented:
Pushed patch again. Last time forgot to mention Reported-by. Mike Snitzer noted this issue and suggested to disable online discards by default.

snitm commented:
Thanks Vivek.  LGTM.

runcom commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1366,10 +1366,7 @@ func (devices *DeviceSet) MountDevice(hash, path, mountLabel string) error {
-	if err != nil && err == syscall.EINVAL {
-		err = syscall.Mount(info.DevName(), path, fstype, flags, options)
-	}
+	err = syscall.Mount(info.DevName(), path, fstype, flags, options)
```
nit but this may be compacted like
```
if err := syscall.Mount(info.DevName(), path, fstype, flags, options); err != nil {
    return fmt.Errorf("Error mounting '%s' on '%s': %s", info.DevName(), path, err)
}
```

rhvgoyal replied:
Yep, will change it.

crosbymichael commented:
LGTM

rhvgoyal committed:
devmapper: Disable mount option "discard" by default

Right now devicemapper mounts thin device using online discards by default
and passes mount option "discard". Generally people discourage usage of
online discards as they can be a drain on performance. Instead it is 
recommended to use fstrim once in a while to reclaim the space.

In case of  containers, we recommend to keep data volumes separate. So
there might not be lot of rm, unlink operations going on and there might
not be lot of space being freed by containers. So it might not matter
much if we don't reclaim that free space in pool.

User can still pass mount option explicitly using dm.mountopt=discard to
enable discards if they would like to.

So this is more like setting the containers by default for better performance
instead of better space efficiency in pool. And user can change the behavior
if they don't like default behavior.

Reported-by: Mike Snitzer <snitzer@redhat.com>
Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
Pused patch again to take care of the comment of making it even more compact.

unclejack commented:
LGTM
